sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -281,6 +281,23 @@ case class EnsureRequirements(
     }
   }
 
+  private def ensureOrdering(plan: SparkPlan, distribution: Distribution) = {
+    (plan.outputPartitioning, distribution) match {
+      case (p @ KeyGroupedPartitioning(expressions, _, partitionValues, _),
+        d @ OrderedDistribution(ordering)) if p.satisfies(d) =>
+        val attrs = expressions.flatMap(_.collectLeaves()).map(_.asInstanceOf[Attribute])
+        val partitionOrdering: Ordering[InternalRow] = {
+          RowOrdering.create(ordering, attrs)
+        }
+        // Sort 'commonPartitionValues' and use this mechanism to ensure BatchScan's
```

Comment:
curious why we need to do this?

Reply:
yea its been a little while, but iiuc its the easiest way to implement it, because internally BatchScanExec bases the ordering of its outputPartitions based on this data structure.

https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala#L127 and https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala#L231

It may also be possible to add another path to sort the partitions in the BatchScanExec, but it would change a lot more code.

Reply:
I see, thanks. I don't have any better ideas on this. For now, it looks fine. Maybe we could think about how to refactor this code better - right now the handlings of`KeyGroupedPartitioning` are scattered around in different places.